File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/optimize/score.mjs
import { loadAssets } from "../lint/scorer.mjs";

function tokenize(text) {
  return (text ?? "")
    .toLowerCase()
    .split(/\W+/)
    .filter(Boolean);
}

function contentWords(text, stop) {
  return tokenize(text).filter((t) => !stop.has(t));
}

function matches(query, descWords, stop) {
  const qw = contentWords(query, stop);
  if (qw.length === 0) return false;
  return qw.some((w) => descWords.has(w));
}

export function hitRate(description, queries) {
  const { stop } = loadAssets();
  const descWords = new Set(contentWords(description, stop));
  const positive = queries.positive ?? [];
  const negative = queries.negative ?? [];
  const positive_hits = positive.filter((q) => matches(q, descWords, stop)).length;
  const false_positives = negative.filter((q) => matches(q, descWords, stop)).length;
  const positive_total = positive.length;
  const false_total = negative.length;
  const tp_rate = positive_total ? positive_hits / positive_total : 0;
  const fp_rate = false_total ? false_positives / false_total : 0;
  const score = Math.max(0, Math.min(100, Math.round((tp_rate - fp_rate) * 100)));
  return {
    positive_hits,
    positive_total,
    false_positives,
    false_total,
    score,
  };
}

File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/optimize/score.py
from __future__ import annotations

import re

from ..lint.scorer import load_assets


def _tokenize(text: str) -> list[str]:
    return [t for t in re.split(r"\W+", (text or "").lower()) if t]


def _content_words(text: str, stop: set) -> list[str]:
    return [t for t in _tokenize(text) if t not in stop]


def _matches(query: str, desc_words: set, stop: set) -> bool:
    qw = _content_words(query, stop)
    if not qw:
        return False
    return any(w in desc_words for w in qw)


def hit_rate(description: str, queries: dict) -> dict:
    assets = load_assets()
    stop = assets["stop"]
    desc_words = set(_content_words(description or "", stop))
    positive = queries.get("positive", []) or []
    negative = queries.get("negative", []) or []
    positive_hits = sum(1 for q in positive if _matches(q, desc_words, stop))
    false_positives = sum(1 for q in negative if _matches(q, desc_words, stop))
    positive_total = len(positive)
    false_total = len(negative)
    tp_rate = (positive_hits / positive_total) if positive_total else 0
    fp_rate = (false_positives / false_total) if false_total else 0
    raw = (tp_rate - fp_rate) * 100
    # round-half-up (match JS Math.round) — avoids Python's banker's rounding
    rounded = int(raw + 0.5) if raw >= 0 else -int((-raw) + 0.5)
    score = max(0, min(100, rounded))
    return {
        "positive_hits": positive_hits,
        "positive_total": positive_total,
        "false_positives": false_positives,
        "false_total": false_total,
        "score": score,
    }
